### PR TITLE
GH-114695: Add `sys._clear_internal_caches`

### PR DESCRIPTION
**Pull request by brandtbucher**
This new function immediately releases all tier-two related references and memory, which is useful for refleak-hunting.

Currently, invalid executors need to be hit again in order to be cleared (which may never happen), and there is no way to free the code object's executor array. This PR adds a (weak) pointer from executors back to the code object they're installed in, which allows them to immediately remove themselves upon invalidation. Code objects can clear this pointer if the executor outlives them.

This PR also changes the linked list to contain every valid executor, and no invalid ones. That means that the old `linked` member is redundant with the `valid` field, so it's removed.

* Issue: gh-114695

----
📚 Documentation preview 📚: https://cpython-previews--115152.org.readthedocs.build/

### CODE REVIEW

File: Include/cpython/optimizer.h [hunk 1]
```diff
@@ -24,9 +24,10 @@ typedef struct {
     uint8_t opcode;
     uint8_t oparg;
     uint8_t valid;
-    uint8_t linked;
     _PyBloomFilter bloom;
     _PyExecutorLinkListNode links;
+    PyCodeObject *code;  // Weak (NULL if no corresponding ENTER_EXECUTOR).
+    int index;           // Index of ENTER_EXECUTOR in the above code object.
```
Comment (markshannon) [inline Include/cpython/optimizer.h]: Can you move this above the bloom filter, to avoid holes?

File: Python/optimizer.c [hunk 2]
```diff
@@ -1151,15 +1158,14 @@ _Py_Executors_InvalidateDependency(PyInterpreterState *interp, void *obj)
...
-    /* Walk the list of executors */
-    for (_PyExecutorObject *exec = interp->executor_list_head; exec != NULL;) {
-        assert(exec->vm_data.valid);
-        _PyExecutorObject *next = exec->vm_data.links.next;
-        exec->vm_data.links.next = NULL;
-        exec->vm_data.links.previous = NULL;
-        exec->vm_data.valid = false;
-        exec->vm_data.linked = false;
-        exec = next;
+    _PyExecutorObject *executor;
```
Comment (markshannon) [inline Python/optimizer.c]: What's the reason for moving the declaration out of the loop?

**Review by markshannon — COMMENTED** [review]
The code looks good, but I have a couple of concerns.

1. I don't like exposing low-level implementation details like executors though, so could we rename `sys._clear_executors()`?
Maybe merge it with `sys._clear_type_cache()` into something like `sys._clear_internal_caches()`? 
That way we can add other caches, jitted code, etc. and clear them all without changing the API.

2. We might want to invalidate a lot of executors in a stop-the-world event. Merely setting the `valid` flag to zero (as we do now) is fine, but freeing lots of object may not be. Can we go back to invalidating executors not doing much work, and leaving clean up to later?

File: Python/optimizer.c (hunk 2 above)
Comment (brandtbucher) [reply Python/optimizer.c]: This is a while loop now.

Would you prefer to leave it as a for loop? It's a tiny bit more repetitive, but maybe easier to follow.

Comment (markshannon) [reply Python/optimizer.c]: The for loop is nice as it limits the scope of the iteration variable. It doesn't matter much.

**Review by markshannon — APPROVED** [review]

**Review by offensive-vk — APPROVED** [review]

**Pull request merged**